### PR TITLE
Eos 24900 : Cluster stop message handling

### PR DESCRIPTION
**Pull request by mukhtar-inamdar**
# Problem Statement
- Clustor stop monitor will receive the message "start_cluster_shutdown" on message bus after that HA needs to stop publishing messages

# Design
-  cluster stop monitor set key in confstore
- K8s monitor will read it and stop publishing messages
- K8s monitor once recive SIGTERM will delete this key form confstore

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM
- [EOS-24900.log](https://github.com/Seagate/cortx-ha/files/7829831/EOS-24900.log)

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [x] Is there a change in filename/package/module or signature? [Y/N]: **N**
- [x] If yes for the above point, is a notification sent to all other cortx components? [Y/N] **N**
- [x] Side effects on other features (deployment/upgrade)? [Y/N] **N**
- [x] Dependencies on other component(s)? [Y/N] **N**
-     If yes for the above point, post a link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide
https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/730890496/Fault+tolerance+LLD

### CODE REVIEW

File: ha/k8s_setup/ha_start.py [hunk 1]
```diff
@@ -30,11 +32,29 @@ def usage(prog: str):
             f"--services health_monitor, fault_tolerance, k8s_monitor\n"
             f"--config   Config URL\n")
 
+def handle_signal(signum, frame):
+    """
+    Sif the process is started already then forward the received signal
+    """
+    print(f"Signal {signum} at frame {frame} received.")
+    print(f"Sending {signum} to driver process pid: {driver_process.pid}.")
```
Comment (akash2144) [inline ha/k8s_setup/ha_start.py:40]: Please remove print statement from whole PR.

File: ha/k8s_setup/ha_start.py [hunk 2]
```diff
@@ -30,11 +32,29 @@ def usage(prog: str):
...
+    driver_process.send_signal(signum)
+
 try:
     if args.services not in service_entry_mapping:
         print('Please provide a valid service name')
         usage('ha_start')
         sys.exit(22)
+    # aetting Signal handlers for all standered signals
```
Comment (Madhura-08) [inline ha/k8s_setup/ha_start.py:48]: Why there is a need to handle all the standard signals?

File: ha/k8s_setup/ha_start.py [hunk 3]
```diff
@@ -30,11 +32,29 @@ def usage(prog: str):
...
+    signal.signal(signal.SIGABRT, handle_signal)
+    signal.signal(signal.SIGFPE, handle_signal)
+    signal.signal(signal.SIGILL, handle_signal)
+    signal.signal(signal.SIGINT, handle_signal)
+    signal.signal(signal.SIGSEGV, handle_signal)
+    signal.signal(signal.SIGTERM, handle_signal)
+
     driver_process = Popen(['/usr/bin/python3', service_entry_mapping[args.services]], shell=False, stdout=PIPE, stderr=PIPE) # nosec
```
Comment (Madhura-08) [inline ha/k8s_setup/ha_start.py:56]: Don't you think this driver_process var will be overwritten? So, lastly, SIGTERM will be sent only to that thread which was lastly executed or started. Right?

Comment (mukhtar-inamdar) [reply ha/k8s_setup/ha_start.py:56]: this driver process starts only 1 process not all at a time and in 1 container.
so there is no reason to overwrite.

File: ha/k8s_setup/ha_start.py (hunk 2 above)
Comment (mukhtar-inamdar) [reply ha/k8s_setup/ha_start.py:48]: just to get all the signals from Kubernetes, don't know which signal will be sent by  Kubernetes in which scenarios.
as our process is a child process, the not exactly main process of the container and Kubernetes signal is sent to only the main process of the container, not a child process so I want to make sure all signals are getting transfer to our process which suppose be the main process of container.

File: ha/k8s_setup/ha_start.py (hunk 1 above)
Comment (mukhtar-inamdar) [reply ha/k8s_setup/ha_start.py]: ok, thanks,
I have sys.stdout.write.

File: ha/k8s_setup/ha_start.py (hunk 2 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_start.py:48]: Ok. Thanks!

File: ha/k8s_setup/ha_start.py (hunk 3 above)
Comment (Madhura-08) [reply ha/k8s_setup/ha_start.py:56]: Got it. Thanks!

**Review by Madhura-08 — APPROVED** [review]

**Comment by shaileshsvc** [comment]
Complete the review checklist with appropriate responses.

**Comment by shaileshsvc** [comment]
As of now I am seeing codacy checks pending (?), please check for positive report when it completes.

File: ha/k8s_setup/ha_start.py (hunk 1 above)
Comment (akash2144) [reply ha/k8s_setup/ha_start.py]: Resolved.

**Review by akash2144 — APPROVED** [review]

File: ha/test/integration/test_sigterm_to_ha_services.py [hunk 4]
```diff
@@ -35,70 +40,105 @@ def start_all_services(self):
...
+        """
+        services = [service for service in self.services]
+
+        # filter out the running services
+        print("Check running services.")
+        services[:] = filterfalse(lambda service : service.poll() != None, services)
+        print(f"Running service: {[service.args for service in services]}")
+        self.assertTrue(len(services) == len(self.services), "Not all HA services are running.")
+
+    def check_for_stop(self, wait_time=30) -> int:
```
Comment (shaileshsvc) [inline ha/test/integration/test_sigterm_to_ha_services.py:108]: The default of wait_time is fine for now. Eventually can be initialized in miniprovisioning (need to be made available by provisioning, picked up from services yaml / conf global settings of grace period) in ha confstore and used.

Comment (shaileshsvc) [reply ha/test/integration/test_sigterm_to_ha_services.py:108]: Please resolve and close this comment if it is not present in current cluster.conf.

**Review by shaileshsvc — APPROVED** [review]
Approved

**Comment by shaileshsvc** [comment]
Please record the unit test results in PR and/or Jira

File: ha/monitor/k8s/object_monitor.py [hunk 5]
```diff
@@ -82,6 +88,31 @@ def handle_sigterm(self, k8s_watch: watch.Watch, k8s_watch_stream):
...
+        """
+        If publish is enabled, i.e. Yet the cluster stop message has not been received.
+        then publish the alert for Fault Tolerance Monitor
+        """
+        if self.is_publish_enable():
+            # Write to message bus
+            Log.info(f"{self._object}_monitor sending alert on message bus {alert.to_dict()}")
+            self._producer.publish(str(alert.to_dict()))
+        else:
+            Log.debug(f"{self._object}_monitor skipping publish for alert: {str(alert.to_dict())}.")
```
Comment (ArchanaLimaye) [inline ha/monitor/k8s/object_monitor.py:114]: Please modify the log to indicate publish skipped since cluster stop received

File: ha/const.py [hunk 6]
```diff
@@ -206,6 +206,9 @@
 RESOURCE_TYPE = "resource_type"
 STONITH_AUTH_TYPE = 'PASSWORD'
 logger_utility_iec_cmd="logger -i -p local3.err"
+CLUSTER_STOP_KEY = "cluster_stop_key"
+CLUSTER_STOP_VAL_ENABLE = "1"
+CLUSTER_STOP_VAL_DISABLE = "0"
```
Comment (ArchanaLimaye) [inline ha/const.py:211]: why do we need ENABLE / DISABLE ? 
We just need to see if key exists or not.

File: ha/k8s_setup/ha_start.py [hunk 7]
```diff
@@ -30,11 +32,29 @@ def usage(prog: str):
             f"--services health_monitor, fault_tolerance, k8s_monitor\n"
             f"--config   Config URL\n")
 
+def handle_signal(signum, frame):
+    """
+    Sif the process is started already then forward the received signal
```
Comment (ArchanaLimaye) [inline ha/k8s_setup/ha_start.py:37]: typo ?

File: ha/const.py (hunk 6 above)
Comment (mukhtar-inamdar) [reply ha/const.py:211]: yes, we can just check key is exist, but some value needs to put for the key.
and if tomorrow we need to implement cancel we can use disable, so I have added it as a confirmation check.
and this if will hit only once in life.

Comment (mukhtar-inamdar) [reply ha/const.py:211]: As discussed keeping enable value only (as a default value but not checking anywhere) and removing disable key as not using anywhere.

File: ha/k8s_setup/ha_start.py (hunk 7 above)
Comment (mukhtar-inamdar) [reply ha/k8s_setup/ha_start.py:37]: corrected, Done

File: ha/monitor/k8s/object_monitor.py (hunk 5 above)
Comment (mukhtar-inamdar) [reply ha/monitor/k8s/object_monitor.py:114]: done

**Review by ArchanaLimaye — APPROVED** [review]

**Pull request merged**